examples/shopping_cart.rs

```diff
@@ -0,0 +1,72 @@
+#![feature(type_alias_enum_variants)]
+
+#[macro_use]
+extern crate mirai_annotations;
+
+pub mod foreign_contracts;
+
+struct Cart {
+    items: Vec<i32>
+}
+
+impl Cart {
+    fn new() -> Cart {
+        Cart {
+            items: Vec::new(),
+        }
+    }
+
+    fn add_item(&mut self, value: i32) {
+        precondition!(value > 0);
+        let old_len = self.items.len();
+        self.items.push(value);
```

Comment:
Although there is no need to say anything in the code here, interesting things are happening in MIRAI: push has a precondition that must be satisfied by this call site. To do so, MIRAI infers a precondition for add_item. The calls to add_item are then checked not just for the explicit precondition, but also for the inferred precondition.